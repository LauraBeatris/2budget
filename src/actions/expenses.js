import uuid from "uuid";

//ADD_EXPENSE
export const addExpense = ({
  description = "",
  note = "",
  amount = 0,
  createdAt = 0
} = {}) => {
  return {
    type: "ADD_EXPENSE",
    expense: {
      // Setting a id
      id: uuid(),
      description,
      note,
      amount,
      createdAt
    }
  };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

//EDIT_EXPENSE
export const editExpense = (id, update) => ({
  type: "EDIT_EXPENSE",
  id,
  update
});
