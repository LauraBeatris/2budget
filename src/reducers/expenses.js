const expensesReducerDefaultState = [];

const expensesReducer = (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(expense => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.update // overriden the expense items with the action updates icons
          };
        } else {
          return expense;
        }
      });
    default:
      return state;
  }
};

export default expensesReducer;
