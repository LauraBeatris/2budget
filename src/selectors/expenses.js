import moment from "moment";

// Get visible expenses
export const getVisibleExpenses = (
  expenses,
  { text, sortBy, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return b.createdAt < a.createdAt ? -1 : 1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? -1 : 1;
      } else {
        return null;
      }
    });
};

export const getExpenseTotal = expenses => {
  if (expenses.length === 0) {
    return 0;
  } else if (expenses.length > 1) {
    const amounts = expenses.map(expense => expense.amount);
    return amounts.reduce((sum, value) => sum + value, 1);
  } else {
    return expenses[0].amount;
  }
};
