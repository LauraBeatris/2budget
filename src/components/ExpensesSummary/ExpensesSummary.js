import React from 'react'
import moment from 'moment'
import { getVisibleExpenses } from '../../selectors/expenses';
import {getExpenseTotal} from "../../selectors/expenses";

import { connect } from 'react-redux'
import numeral from 'numeral'

export const ExpensesSummary = ({expensesQuantity, expensesTotal}) => (
    <div className="summary">
    <p className="moment">{moment().format("MMMM YYYY")}</p>
    <p className="quantity">{expensesQuantity > 1 ? expensesQuantity + ' Items' : expensesQuantity + ' Item'}</p>
   
    <h1 className="total-amount">{numeral(expensesTotal / 100).format('$0,0.00')}</h1>
    <h2>Total Amount</h2>
  </div>
)

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters)

    return {
      expensesQuantity: visibleExpenses.length,
      expensesTotal: getExpenseTotal(visibleExpenses)
    }
}
export default connect(mapStateToProps)(ExpensesSummary)