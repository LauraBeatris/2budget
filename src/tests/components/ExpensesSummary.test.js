import React from 'react';
import { shallow } from 'enzyme'
import { getVisibleExpenses } from '../../selectors/expenses';
import {getExpenseTotal} from "../../selectors/expenses"; 
import {ExpensesSummary} from "../../components/ExpensesSummary/ExpensesSummary";
import expenses from "../fixtures/expenses"


test('should correctly render ExpensesSummary with 1 expense', () => {
    const expensesTotal = getExpenseTotal([expenses[0]])
    const wrapper = shallow(<ExpensesSummary expenses={{expensesTotal, expensesQuantity: 1}}/>)
    expect(wrapper).toMatchSnapshot()

    
})

test('should correctly render ExpensesSummary with more than 1 expense', () => {
    const expensesTotal = getExpenseTotal([expenses])
    const wrapper = shallow(<ExpensesSummary expenses={{expensesTotal, expensesQuantity: 2}}/>)

    expect(wrapper).toMatchSnapshot()  
})
