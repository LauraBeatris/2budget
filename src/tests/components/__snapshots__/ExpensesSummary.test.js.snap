// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with 1 expense 1`] = `
<div
  className="summary"
>
  <p
    className="moment"
  >
    December 1969
  </p>
  <p
    className="quantity"
  >
    undefined Item
  </p>
  <h1
    className="total-amount"
  >
    $0.00
  </h1>
  <h2>
    Total Amount
  </h2>
</div>
`;

exports[`should correctly render ExpensesSummary with more than 1 expense 1`] = `
<div
  className="summary"
>
  <p
    className="moment"
  >
    December 1969
  </p>
  <p
    className="quantity"
  >
    undefined Item
  </p>
  <h1
    className="total-amount"
  >
    $0.00
  </h1>
  <h2>
    Total Amount
  </h2>
</div>
`;
